src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -272,65 +319,105 @@ fn save_file(
-      })
-      .and_then(move |stats| {
-        let mut dirs = Vec::new();
-        let mut files = Vec::new();
-        for stat in stats.into_iter() {
-          match stat {
-            fs::Stat::Dir(dir) => dirs.push(dir),
-            fs::Stat::File(file) => files.push(file),
-            fs::Stat::Link(fs::Link(path)) => {
-              let err: BoxFuture<_, _> =
-                futures::future::err(format!("Don't yet know how to handle symlinks: {:?}", path))
-                  .to_boxed();
-              return err;
-            }
-          }
-        }
+  let mut file_futures: Vec<BoxFuture<bazel_protos::remote_execution::FileNode, String>> =
+    Vec::new();
+  let mut dir_futures: Vec<BoxFuture<bazel_protos::remote_execution::DirectoryNode, String>> =
+    Vec::new();
 
-        let store_inner_copy = store_copy.clone();
-        let posix_fs_copy = posix_fs.clone();
-        let dir_futures: Vec<_> = dirs
-          .into_iter()
-          .map(move |dir| {
-            let dir = Arc::new(dir);
-            save_directory(store_inner_copy.clone(), posix_fs_copy.clone(), dir.clone())
-              .and_then(move |(fingerprint, proto_size_bytes)| {
-                let mut dir_node = bazel_protos::remote_execution::DirectoryNode::new();
-                dir_node.set_name(file_name_as_utf8(&dir.0)?);
-                dir_node.set_digest(fingerprint_to_digest(&fingerprint, proto_size_bytes as i64));
-                Ok(dir_node)
-              })
-          })
-          .collect();
+  for (first_component, group) in
+    &path_stats.into_iter().group_by(|s| {
+      s.path().components().next().unwrap().as_os_str().to_owned()
+    })
+  {
+    let path_group: Vec<fs::PathStat> = group.collect();
+    if path_group.len() == 1 && path_group.get(0).unwrap().path().components().count() == 1 {
+      // Exactly one entry with exactly one component indicates either a file in this directory, or
+      // an empty directory.
+      // If the child is a non-empty directory, or a file therein, there must be multiple PathStats
+      // with that prefix component, and we will handle that in the recursive save_directory call.
 
-        let file_futures: Vec<_> = files
-          .into_iter()
-          .map(move |file| {
-            file_to_file_node(store_copy.clone(), posix_fs.clone(), file)
-          })
-          .collect();
+      match path_group.get(0).unwrap() {
+        &fs::PathStat::File { ref stat, .. } => {
+          let stat = stat.clone();
+          file_futures.push(
+            save_file(store.clone(), posix_fs.clone(), stat.clone())
+              .and_then(move |(fingerprint, size_bytes)| {
+                let mut file_node = bazel_protos::remote_execution::FileNode::new();
+                file_node.set_name(osstring_as_utf8(&first_component).unwrap());
+                file_node.set_digest(fingerprint_to_digest(&fingerprint, size_bytes as i64));
+                file_node.set_is_executable(stat.is_executable);
```

Comment:
If you capture `is_executable` in a local, you can move that into the closure rather than cloning the `stat`:
```
let is_executable = stat.is_executable;
...
  file_node.set_is_executable(is_executable);
```

Reply:
Done